contracts/src/registry/NameWrapperRegistry.sol

```diff
@@ -0,0 +1,247 @@
+        uint64 oldExpiry = _extractExpiry(oldFlags);
+        if (oldExpiry >= block.timestamp) {
+            revert NameAlreadyRegistered(label);
+        }
+
+        // if there is a previous owner, burn the token
+        address previousOwner = super.ownerOf(tokenId);
+        if (previousOwner != address(0)) {
+            _burn(previousOwner, tokenId, 1);
+        }
+
+        _mint(owner, tokenId, 1, "");
+        datastore.setSubregistry(tokenId, address(registry), flags);
+
+        // registrar has some permissions by default
+        bytes32 tokenIdContext = _tokenIdContext(tokenId);
+        _grantRole(tokenIdContext, RENEW_ROLE, msg.sender);
+        _grantRole(tokenIdContext, EMANCIPATOR_ROLE, msg.sender);
+
+        emit NewSubname(label);
+        return tokenId;
+    }
+
+    function renew(uint256 tokenId, uint64 expires) public onlyRole(_tokenIdContext(tokenId), RENEW_ROLE) {
+        address sender = _msgSender();
+
+        (address subregistry, uint96 flags) = datastore.getSubregistry(tokenId);
+        uint64 oldExpiration = _extractExpiry(flags);
+        if (oldExpiration < block.timestamp) {
+            revert NameExpired(tokenId);
+        }
+        if (expires < oldExpiration) {
+            revert CannotReduceExpiration(oldExpiration, expires);
+        }
+        datastore.setSubregistry(tokenId, subregistry, (flags & FLAGS_MASK) | (uint96(expires) << 32));
+
+        emit NameRenewed(tokenId, expires, sender);
+    }
+
+    /** 
+     * @dev Emancipate the owner of the token. 
+     * 
+     * See v1 fuse: PARENT_CANNOT_CONTROL
+     */
+    function emancipate(uint256 tokenId) public onlyRole(_tokenIdContext(tokenId), EMANCIPATOR_ROLE) {
+        bytes32 tokenIdContext = _tokenIdContext(tokenId);
+        _grantRole(tokenIdContext, EMANCIPATED_OWNER_ROLE, ownerOf(tokenId));
+        renounceRole(tokenIdContext, EMANCIPATOR_ROLE, msg.sender); // remove the emancipator role since we can only do this once.
+    }
+    
+    /**
+     * @dev Allow the owner to renew the name.
+     * 
+     * See v1 fuse: CAN_EXTEND_EXPIRY
+     */
+    function allowOwnerToRenew(uint256 tokenId) public onlyRole(_tokenIdContext(tokenId), EMANCIPATOR_ROLE) {
+        bytes32 tokenIdContext = _tokenIdContext(tokenId);
+        _grantRole(tokenIdContext, RENEW_ROLE, ownerOf(tokenId));
+    }
+
+
+    /**
+     * @dev Prevent the name from being renewed.
+     * 
+     * See v1 fuse: CANNOT_SET_TTL
+     */
+    function lockRenewals(uint256 tokenId) public onlyRole(_tokenIdContext(tokenId), EMANCIPATED_OWNER_ROLE) {
+        bytes32 tokenIdContext = _tokenIdContext(tokenId);
+        _revokeRoleAssignments(tokenIdContext, RENEW_ROLE);
+    }
+
+    /**
+     * @dev Prevent the name from being transferred.
+     * 
+     * See v1 fuse: CANNOT_TRANSFER
+     */
+    function lockTransfers(uint256 tokenId) public onlyRole(_tokenIdContext(tokenId), EMANCIPATED_OWNER_ROLE) {
```

Comment:
Are these functions presently only available on emancipated names?

Reply:
I'm not actually sure we need this. The sole right that owning the NFT gives you is the right to transfer it, so if you want to block transfers, you could simply transfer the name to a burn address. All other functionality would continue to work fine!

Reply:
I was just replicate fuses, but sure, will remove it.